File: Hadoop_Map_reduce_code/join/src/main/java/hadoop/join/UsersZipcodes.java
package hadoop.join;

import java.net.URI;
import java.util.HashMap;

import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.fs.FileSystem;
import org.apache.hadoop.fs.Path;
import org.apache.hadoop.io.Text;
import org.apache.hadoop.mapreduce.Job;
//import org.apache.hadoop.mapreduce.lib.input.FileInputFormat;
import org.apache.hadoop.mapreduce.lib.input.MultipleInputs;
import org.apache.hadoop.mapreduce.lib.input.TextInputFormat;
import org.apache.hadoop.mapreduce.lib.output.FileOutputFormat;



public class UsersZipcodes {

	public static void main(String[] args) throws Exception{
		
		Configuration conf = new Configuration();
		 
		HashMap<String, Integer> mapUsers = new HashMap<>(); 
		mapUsers.put("userid",0);
		mapUsers.put("age",1);
		mapUsers.put("gender",2);
		mapUsers.put("occupation",3);
		mapUsers.put("zipcode",4);
		
		
		HashMap<String, Integer> mapZipcodes = new HashMap<>(); 
		mapZipcodes.put("zipcode",0);
		mapZipcodes.put("zipcodetype",1);
		mapZipcodes.put("city",2);
		mapZipcodes.put("state",3);
		
		String attribute="DEFAULT";
		String val = "DEFAULT";
		if(args.length==2)
		{
			attribute=args[0];
			val=args[1];
		}

		if(mapUsers.containsKey(attribute))
		{
			conf.set("table1","users");
			conf.set("attributeIndex1", String.valueOf(mapUsers.get(attribute)));
			conf.set("val1", val);
		}
		
		if(mapZipcodes.containsKey(attribute))
		{
			conf.set("table2","zipcodes");
			conf.set("attributeIndex2", String.valueOf(mapZipcodes.get(attribute)));
			conf.set("val2", val);
		}
		
		Job job = new Job(conf, "Users Zipcodes");
		 job.setJarByClass(UsersZipcodes.class);
		 job.setReducerClass(UsersZipcodesReducer.class);
		 job.setOutputKeyClass(Text.class);
		 job.setOutputValueClass(Text.class);
		 
		 
		 
		 MultipleInputs.addInputPath(job, new Path("/cloud/input/users.csv"),TextInputFormat.class, UsersMapper.class);
		 MultipleInputs.addInputPath(job, new Path("/cloud/input/zipcodes.csv"),TextInputFormat.class, ZipcodesMapper.class);
		 Path outputPath = new Path("/cloud/output/userszipcodes");
		  
		 FileOutputFormat.setOutputPath(job, outputPath);
		 
		 FileSystem hdfs = FileSystem.get(URI.create("hdfs://localhost:9000"),conf);
			
			// delete existing directory
			if (hdfs.exists(outputPath)) {
			    hdfs.delete(outputPath, true);
			}
		 
		 
		 
		 System.exit(job.waitForCompletion(true) ? 0 : 1);
		 	
	}

}

File: spark_scripts/userszipcodes.py
from pyspark import SparkContext, SparkConf
import sys
import time

start_time = time.time()

conf = SparkConf().setAppName("userszipcodes")
sc = SparkContext(conf = conf)

URI = sc._gateway.jvm.java.net.URI
Path = sc._gateway.jvm.org.apache.hadoop.fs.Path
FileSystem = sc._gateway.jvm.org.apache.hadoop.fs.FileSystem
fs = FileSystem.get(URI("hdfs://localhost:9000"), sc._jsc.hadoopConfiguration())


usersRDD = sc.textFile("hdfs://localhost:9000/cloud/input/users.csv")
usersRDDMap = usersRDD.map(lambda rec: (rec.split(",")[4], rec))

zipcodesRDD = sc.textFile("hdfs://localhost:9000/cloud/input/zipcodes.csv")
zipcodesRDDMap = zipcodesRDD.map(lambda rec: (rec.split(",")[0], rec))

userszipcodesRDDMap=usersRDDMap.join(zipcodesRDDMap)
userszipcodesRDDMapF = userszipcodesRDDMap

mapUsers =	{
  "userid": 0,
  "age": 1,
  "gender": 2,
  "occupation": 3,
  "zipcode": 4,
}

mapZipcodes = {
  "zipcode": 0,
  "zipcodetype": 1,
  "city": 2,
  "state": 3,
}

def main():
  arguments = sys.argv[1:]
  count = len(arguments)
  if count == 2:
    attribute = arguments[0]
    val = arguments[1]
    attributeIndexUsers = mapUsers.get(attribute)
    attributeIndexZipcodes = mapZipcodes.get(attribute)
    if attributeIndexUsers != None:
      if attribute == 'userid' or attribute == 'age':
        userszipcodesRDDMapF = userszipcodesRDDMap.filter(lambda row: int(row[1][0].split(",")[attributeIndexUsers]) == int(val))
      else:
        userszipcodesRDDMapF = userszipcodesRDDMap.filter(lambda row: (val in row[1][0].split(",")[attributeIndexUsers]))
    elif attributeIndexZipcodes != None:
      userszipcodesRDDMapF = userszipcodesRDDMap.filter(lambda row: (val in row[1][1].split(",")[attributeIndexZipcodes]))
    else:
      userszipcodesRDDMapF = userszipcodesRDDMap
  else:
    userszipcodesRDDMapF = userszipcodesRDDMap
  
  
  fs.delete(Path('/cloud/output/spark/userszipcodes') ,True)
  userszipcodesRDDMapF.coalesce(1, shuffle = True).saveAsTextFile("hdfs://localhost:9000/cloud/output/spark/userszipcodes")


  execution_time = time.time() - start_time
  f = open("extime.txt", "w")
  f.write(str(execution_time))
  f.close()

if __name__== "__main__":
  main()
